test/test/multi_threaded_map.cpp

```diff
@@ -0,0 +1,313 @@
+        REQUIRE(totalIncrements == (size - 1));
+    }
+
+    // HasCurrent / GetMany
+    {
+        auto itr = map.First();
+        unique_thread threads[2];
+        int totals[std::size(threads)] = {};
+        for (int i = 0; i < std::size(threads); ++i)
+        {
+            threads[i] = unique_thread([&itr, &totals, i]
+            {
+                IKeyValuePair<int, T> vals[10];
+                while (itr.HasCurrent())
+                {
+                    // Unlike 'Current', 'GetMany' _is_ atomic in regards to read+increment
+                    auto len = itr.GetMany(vals);
+                    totals[i] += std::accumulate(vals, vals + len, 0, [](int curr, auto const& next) { return curr + next.Key(); });
+                }
+            });
+        }
+
+        for (auto& t : threads)
+        {
+            t.join();
+        }
+
+        // sum(i = 1 -> N){i} = N * (N + 1) / 2
+        auto total = std::accumulate(std::begin(totals), std::end(totals), 0);
+        REQUIRE(total == (size * (size - 1) / 2));
+    }
+}
+
+template <typename T>
+static void test_multi_writer(IMap<int, T> const& map)
+{
+    // Large enough that several threads should be executing concurrently
+    static constexpr uint32_t size = 10000;
+    static constexpr size_t threadCount = 8;
+
+    // Insert
+    unique_thread threads[threadCount];
+    for (int i = 0; i < threadCount; ++i)
+    {
+        threads[i] = unique_thread([&map, i]
+        {
+            auto off = i * size;
+            for (int j = 0; j < size; ++j)
+            {
+                map.Insert(j + off, conditional_box<T>(j));
+            }
+        });
+    }
+
+    for (auto& t : threads)
+    {
+        t.join();
+    }
+
+    REQUIRE(map.Size() == (size * threadCount));
+
+    // sum(i = 1 -> N){i} = N * (N + 1) / 2
+    auto sum = std::accumulate(begin(map), end(map), 0, [](int curr, auto&& pair) { return curr + conditional_unbox(pair.Value()); });
+    REQUIRE(sum == ((threadCount * (size - 1) * size) / 2));
+}
+
+struct exclusive_map :
+    map_base<exclusive_map, int, int>,
+    implements<exclusive_map, IMap<int, int>, IMapView<int, int>, IIterable<IKeyValuePair<int, int>>>
+{
+    std::map<int, int> container;
+    mutable slim_mutex mutex;
+
+    auto& get_container() noexcept
+    {
+        return container;
+    }
+
+    auto& get_container() const noexcept
+    {
+        return container;
+    }
+
+    // It is not safe to recursively acquire an SRWLOCK, even in shared mode, however this is unchecked by the SRWLOCK
+    // implementation. Using a vector that only performs exclusive operations is the simplest way to validate that
+    // the implementation does not attempt to recursively acquire the mutex.
+    template <typename Func>
+    auto exclusive_op(Func&& fn) const
+    {
+        slim_lock_guard guard(mutex);
+        return fn();
+    }
+};
+
+static void deadlock_test()
```

Comment:
Need a reentrancy test that validates that objects are not destructed under the lock.
```cpp
// Exercises the case where an object's destructor results in a call
// back into the container it was removed from. This test makes the call directly,
// but in real life, the path is usually more circuitous.
// E.g., destructor of object X releases the final reference to object Y,
// and object Y's destructor tries to remove itself from the same container.

template<typename Container>
struct CallsBackInDestructor : implements<CallsBackInDestructor, IInspectable>
{
    Container m_owner;
    CallsBackInDestructor(Container const& owner) : m_owner(owner) { }
    ~CallsBackInDestructor() { m_owner.Size(); }
};

void reentrancy_test()
{
    auto map = multi_threaded_map<int, IInspectable>();
    map.Insert(1, make<CallsBackInDestructor>(map));
    map.Insert(1, make<CallsBackInDestructor>(map)); // should not deadlock
    map.Remove(1); // should not deadlock

    map.Insert(1, make<CallsBackInDestructor>(map));
    map.Clear(); // should not deadlock
}
```